### DISCUSSION TITLE
Unstructured grids - UGRID from a TDS

### DISCUSSION THREAD

**Discussion by PBrockmann** [Unidata/tds#345] [Q&A]
I have read this proposal and I am wondering if the project https://github.com/ioos/thredds-ugrid has been finally merged into the current Thredds code.

This proposal was a really good initiative.
https://github.com/ioos/thredds-ugrid/blob/master/documents/proposal/2018-0196%20Proposal_Narrative_redacted.pdf
How did this project ended ?

I am looking for an infrastructure able to deliver a WMS for UGRID ressources.
Let me know.

**Comment by haileyajohnson** [comment]
I believe this work is more or less complete but still needs testing... @lesserwhirls can probably provide more information

**Comment by PBrockmann** [comment]
I would be really interested to test it.
There are a couple of files from https://geoport.whoi.edu/ but can provide output from our icosahedral hydrostatic model.

**Comment by PBrockmann** [comment]
Any news ? Having a calendar in mind would be kind.

**Comment by lesserwhirls** [comment]
The general work is complete, and FMRCs can be created using UGRID datasets. The work is not available in the current release (v5.4), but will be in the next release (and should be available in the current v5.5 SNAPSHOTS, available at https://downloads.unidata.ucar.edu/tds/). The WCS service will not work with UGRID datasets, and NCSS currently does not work with UGRID datasets (but will be with a bit more work). An example of a UGRID FMRC can be found here:

https://github.com/Unidata/tds/blob/dda75f8639631db32221405fe017047c1524c4da/tds/src/test/content/thredds/catalogFmrc.xml#L78-L87

The definition of the "ugrid" service is:

https://github.com/Unidata/tds/blob/dda75f8639631db32221405fe017047c1524c4da/tds/src/test/content/thredds/catalog.xml#L24-L32

**Comment by PBrockmann** [comment]
Very good.
Is the WMS will be available for UGRID datasets ?

**Comment by lesserwhirls** [comment]
> Very good.
> 
> Is the WMS will be available for UGRID datasets ?

Yes.

**Comment by PBrockmann** [comment]
Impatient to test it.
Do you have a TDS available for testing ?

**Comment by lesserwhirls** [comment]
I do not - I am no longer an employee at Unidata, and I do not run a TDS from home. Docker might be the easiest way to test.

**Comment by haileyajohnson** [comment]
> Impatient to test it. Do you have a TDS available for testing ?

You can set up your own instance by downloading the latest snapshot [here](https://downloads.unidata.ucar.edu/tds/), or you can see a running instance of the latest snapshot at [thredds-test.unidata.ucar.edu](thredds-test.unidata.ucar.edu) - I don't know if there are any UGRID datasets on it currently though.

**Comment by PBrockmann** [comment]
Ok thank you. I will investigate. Regards

**Comment by PBrockmann** [comment]
Ok tested. Indeed it works with the tested UGRID files I had.

Just how to debug when the GetCapabilities answers 
`Cause: No grids found in underlying NetCDF dataset`

**Comment by PBrockmann** [comment]
It would be interesting to reopen this issue because I think there are some precisions to make. 
Perhaps a GitHub discussion would be better.

I have tried a UGRID found from https://psyplot.github.io/examples/maps/example_ugrid.html
The ugrid_demo.nc. The WMS GetCapacities response seems ok but there are in fact no Layers available.

http://0.0.0.0/thredds/wms/testAll/ugrid_demo.nc?service=WMS&version=1.3.0&request=GetCapabilities
```
<Layer>
<Title>netCDF output from StormFlash2d</Title>
</Layer>
```

Is this file uncorrect from the UGRID convention ?

**Reply by lesserwhirls** [reply]
It might be helpful to check out the ncml used to modify the example file used by some of our tests:

https://github.com/Unidata/tds/blob/main/tds-ugrid/src/test/resources/cases/fvcom/fvcom_delt.ncml

**Reply by PBrockmann** [reply]
Would be nice to collect a set of test files at UGRID format.

https://github.com/sci-wms/sci-wms/blob/master/wms/resources/selfe_ugrid.nc
--> works
![image](https://user-images.githubusercontent.com/5402758/220759537-ea5050fd-1b70-433b-a646-6916802ca09d.png)

But not this one https://psyplot.github.io/examples/maps/example_ugrid.html

What are the ones used to test UGRID from the TDS ?

**Reply by PBrockmann** [reply]
Have found a module to check netcdf files to the UGRID specification. See https://pypi.org/project/ugrid-checks/

When I check the ugrid_demo.nc file, it did not complain and indicates that this file is conform. (file is available from https://psyplot.github.io/examples/maps/example_ugrid.html).

```
ugrid-checker -s ugrid_demo.nc 

File mesh structure
-------------------
Meshes
    "Mesh2"
        node("nMesh2_node")
            coordinates : "Mesh2_node_x", "Mesh2_node_y"
        edge("nMesh2_edge")
            edge_node_connectivity : "Mesh2_edge_nodes"
        face("nMesh2_face")
            face_node_connectivity : "Mesh2_face_nodes"

Mesh Data Variables
    "Mesh2_height"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_bathy"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_m_x"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_m_y"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_u_x"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_u_y"
        mesh : "Mesh2"
        location : "face"
    "Mesh2_level"
        mesh : "Mesh2"
        location : "face"


UGRID conformance checks complete.

No problems found.

Done.
```

When exposed with the last TDS snapshot, the WMS does not display any queryable layer. Do not know how to go further.

Same diagnostic with a very basic UGRID test file with only 2 cells derived from our dynamic model I am working with made of hexaedrons cells.

`$ ncdump dynamico01.nc

```
netcdf dynamico01 {
dimensions:
        faces = 2 ;
        nodes = 12 ;
        six = 6 ;
variables:
        int Mesh ;
                Mesh:cf_role = "mesh_topology" ;
                Mesh:topology_dimension = 2 ;
                Mesh:node_coordinates = "lat_nodes lon_nodes" ;
                Mesh:face_node_connectivity = "face_nodes" ;
        double lon_nodes(nodes) ;
                lon_nodes:standard_name = "longitude" ;
                lon_nodes:long_name = "Longitude of 2D mesh nodes" ;
                lon_nodes:units = "degrees_east" ;
        double lat_nodes(nodes) ;
                lat_nodes:standard_name = "latitude" ;
                lat_nodes:long_name = "Latitude of 2D mesh nodes" ;
                lat_nodes:units = "degrees_north" ;
        int face_nodes(faces, six) ;
                face_nodes:cf_role = "face_node_connectivity" ;
                face_nodes:long_name = "Maps every face to its corner nodes." ;
                face_nodes:start_index = 0 ;
        float aVar(faces) ;
                aVar:mesh = "Mesh" ;
                aVar:location = "face" ;

// global attributes:
                :Conventions = "UGRID-0.9" ;
data:

 Mesh = _ ;

 lon_nodes = 179.722100732443, 179.556350557455, 180, -179.556350557455, 
    -179.556350557455, -179.722100732443, 179.716455159806, 179.539558454989, 
    179.722100732443, -179.722100732443, -179.539558454989, -179.716455159806 ;

 lat_nodes = 60.5859063023705, 60.3262421599437, 60.168182018156, 
    60.3262421599437, 60.3262421599437, 60.5859063023705, 60.9684683053588, 
    60.706675253158, 60.5859063023705, 60.5859063023705, 60.706675253158, 
    60.9684683053587 ;

 face_nodes =
  0, 1, 2, 3, 4, 5,
  6, 7, 8, 9, 10, 11 ;

 aVar = 100, 200 ;
}
```

And here again the file seems to be conform to UGRID convention.

```
ugrid-checker dynamico01.nc

UGRID conformance checks complete.

No problems found.

Done.
```

but same problem: no layer queryable exposed from the WMS service.
Any idea ?

**Reply by PBrockmann** [reply]
Any news ?

**Reply by lesserwhirls** [reply]
No news on my end - I have not had much free time on the weekends to look at it (currently on my lunch break). Part of the issue you might be running into is related to the "Conventions" global attribute. In order for netCDF-Java to recognize the dataset as following the UGRID conventions, the Conventions attribute needs to indicate that it is a subconvention of the CF convention. To do that, the value needs to look something like: `Conventions: CF-1.6/UGRID-1.X` (i.e. both CF and UGRID listed). At least one of the datasets that you point to above as not working only lists UGRID as the convention, so in that case edal-java (ncWMS 2.x) would return a "No grids found in underlying NetCDF dataset" message. The example NcML I link to above shows how to add that attribute without rewriting the file.

**Reply by PBrockmann** [reply]
Cool. Thanks for the hint.
Could you point me the java file that handles the reading of a UGRID file ?

**Reply by PBrockmann** [reply]
Indeed adding the global attribut `Conventions = "CF-1.6/UGRID-0.9"` remove the "No grids found in underlying NetCDF dataset" but I cannot find a queryable layer with this modified file.

```
<Layer>
<Title>No dataset title found.</Title>
</Layer>
```

**Reply by PBrockmann** [reply]
I have tried with the simplest data structure I would like to use. A single hexagon.
The WMS is available but no representation in my application (https://github.com/PBrockmann/mapper).

Here is the cdl:
```
netcdf hexagon {
dimensions:
        faces = 1 ;
        nodes = 6 ;
        six = 6 ;
variables:
        int mesh ;
                mesh:cf_role = "mesh_topology" ;
                mesh:topology_dimension = 2 ;
                mesh:node_coordinates = "mesh_nodes_x mesh_nodes_y" ;
                mesh:face_node_connectivity = "mesh_face_nodes" ;
                mesh:face_dimension = "faces" ;
        double mesh_nodes_x(nodes) ;
                mesh_nodes_x:standard_name = "longitude" ;
                mesh_nodes_x:long_name = "Longitude of 2D mesh nodes" ;
                mesh_nodes_x:units = "degrees_east" ;
        double mesh_nodes_y(nodes) ;
                mesh_nodes_y:standard_name = "latitude" ;
                mesh_nodes_y:long_name = "Latitude of 2D mesh nodes" ;
                mesh_nodes_y:units = "degrees_north" ;
        int mesh_face_nodes(faces, six) ;
                mesh_face_nodes:cf_role = "face_node_connectivity" ;
                mesh_face_nodes:long_name = "Maps every face to its corner nodes." ;
                mesh_face_nodes:start_index = 0 ;
        float aVar(nodes) ;
                aVar:long_name = "a variable" ;
                aVar:mesh = "mesh" ;
                aVar:location = "node" ;

// global attributes:
                :Conventions = "CF-1.6/UGRID-1.X" ;
data:

 mesh = _ ;

 mesh_nodes_x = 20, 30, 40, 50, 40, 30;

 mesh_nodes_y = 20, 30, 30, 20, 10, 10;

 mesh_face_nodes = 0, 1, 2, 3, 4, 5;

 aVar = 20, 30, 30, 20, 10, 10;

}
```

What do I am missing ?

**Reply by lesserwhirls** [reply]
The code that handles interpreting the UGRID convention is located in the `tds-ugrid` subprojtect of the TDS codebase (see [here](https://github.com/Unidata/tds/tree/main/tds-ugrid)). Also, I just noticed that you are dealing with hexahedron cells - this implementation only handles triangular meshes, so that's why no layers are found. I'm surprised there are no error messages in the logs about the face node dimensionality (I'll have to check that out).

**Reply by PBrockmann** [reply]
Ahhh. Only triangles. I understand better.

```
netcdf selfe_ugrid_select1 {
dimensions:
        face = 3 ;
        max_face_nodes = 3 ;
        node = 5 ;
variables:
        int Mesh ;
                Mesh:cf_role = "mesh_topology" ;
                Mesh:long_name = "Topology data of 2D unstructured mesh" ;
                Mesh:topology_dimension = 2 ;
                Mesh:node_coordinates = "node_lon node_lat" ;
                Mesh:face_node_connectivity = "face_nodes" ;
        int face_nodes(face, max_face_nodes) ;
                face_nodes:cf_role = "face_node_connectivity" ;
                face_nodes:long_name = "Maps every face to its corner nodes." ;
                face_nodes:start_index = 1 ;
        double node_lat(node) ;
                node_lat:standard_name = "latitude" ;
                node_lat:long_name = "Latitude of 2D mesh nodes" ;
                node_lat:units = "degrees_north" ;
        double node_lon(node) ;
                node_lon:standard_name = "longitude" ;
                node_lon:long_name = "Longitude of 2D mesh nodes" ;
                node_lon:units = "degrees_east" ;
        float surface_temp(node) ;
                surface_temp:coordinates = "node_lon node_lat" ;
                surface_temp:long_name = "Sea surface temperature" ;
                surface_temp:standard_name = "sea_surface_temperature" ;
                surface_temp:mesh = "Mesh" ;
                surface_temp:location = "node" ;
                surface_temp:units = "degree_Celsius" ;

// global attributes:
                :Conventions = "CF-1.6/UGRID-1.x" ;

data:

 Mesh = _ ;

 face_nodes =
  1, 4, 5,
  1, 5, 2,
  4, 1, 3;

 node_lat = 46.3698348999023, 46.3674011230469,
         46.3713722229004, 46.3680648803711, 46.3661003112793;

 node_lon = -124.104789733887, -124.100074768066,
         -124.110336303711, -124.110466003418, -124.105400085449;

 surface_temp = 200, 100, 0, 50, 80, 150;
}

```

Gives this. 

![image](https://user-images.githubusercontent.com/5402758/223563594-dd8ab807-5d5f-474e-84c5-f3c280fe0a28.png)

But when I remove 2 nodes to have only a triangle, nothing appears.

```
netcdf selfe_ugrid_select1 {
dimensions:
        face = 1 ;
        max_face_nodes = 3 ;
        node = 3 ;
variables:
        int Mesh ;
                Mesh:cf_role = "mesh_topology" ;
                Mesh:long_name = "Topology data of 2D unstructured mesh" ;
                Mesh:topology_dimension = 2 ;
                Mesh:node_coordinates = "node_lon node_lat" ;
                Mesh:face_node_connectivity = "face_nodes" ;
        int face_nodes(face, max_face_nodes) ;
                face_nodes:cf_role = "face_node_connectivity" ;
                face_nodes:long_name = "Maps every face to its corner nodes." ;
                face_nodes:start_index = 1 ;
        double node_lat(node) ;
                node_lat:standard_name = "latitude" ;
                node_lat:long_name = "Latitude of 2D mesh nodes" ;
                node_lat:units = "degrees_north" ;
        double node_lon(node) ;
                node_lon:standard_name = "longitude" ;
                node_lon:long_name = "Longitude of 2D mesh nodes" ;
                node_lon:units = "degrees_east" ;
        float surface_temp(node) ;
                surface_temp:coordinates = "node_lon node_lat" ;
                surface_temp:long_name = "Sea surface temperature" ;
                surface_temp:standard_name = "sea_surface_temperature" ;
                surface_temp:mesh = "Mesh" ;
                surface_temp:location = "node" ;
                surface_temp:units = "degree_Celsius" ;

// global attributes:
                :Conventions = "CF-1.6/UGRID-1.x" ;

data:

 Mesh = _ ;

 face_nodes =
  1, 2, 3;

 node_lat = 46.3698348999023, 46.3674011230469,
         46.3713722229004;

 node_lon = -124.104789733887, -124.100074768066,
         -124.110336303711;

 surface_temp = 200, 100, 0;
}
```

Why ? And last question how to have value on faces, not on nodes ?  An example would be welcomed.

**Reply by lesserwhirls** [reply]
In order for a variable to be recognized as a "gridded" type by netCDF-Java, the variable will need to have a time dimension and time coordinate variable. As for an example, check out the netCDF file and its associated NcML file [here](https://github.com/Unidata/tds/tree/main/tds-ugrid/src/test/resources/cases/fvcom).

**Reply by PBrockmann** [reply]
Not sure to be agree. A variable can be time independant (area of cells, bathymetry, ...).
As a proof, the cdl selfe_ugrid_select1 I have put above is accepted and the WMS is available for mapping.

**Reply by lesserwhirls** [reply]
My apologies, you are correct - I was thinking of restrictions on a different TDS service.

**Reply by drf5n** [reply]
> It would be interesting to reopen this issue because I think there are some precisions to make. Perhaps a GitHub discussion would be better.
> 
> I have tried a UGRID found from https://psyplot.github.io/examples/maps/example_ugrid.html The ugrid_demo.nc. The WMS GetCapacities response seems ok but there are in fact no Layers available.

I hunted down a link to a copy of this sample file:  https://codebase.helmholtz.cloud/psyplot/psy-maps/-/blob/v1.3.1/examples/ugrid_demo.nc

**Comment by PBrockmann** [comment]
Here is what I get from this file that has no time dimension.

```
netcdf selfe_ugrid3 {
dimensions:
        face = 3 ;
        max_face_nodes = 3 ;
        node = 5 ;
variables:
        int Mesh ;
                Mesh:cf_role = "mesh_topology" ;
                Mesh:long_name = "Topology data of 2D unstructured mesh" ;
                Mesh:topology_dimension = 2 ;
                Mesh:node_coordinates = "node_lon node_lat" ;
                Mesh:face_node_connectivity = "face_nodes" ;
        int face_nodes(face, max_face_nodes) ;
                face_nodes:cf_role = "face_node_connectivity" ;
                face_nodes:long_name = "Maps every face to its corner nodes." ;
                face_nodes:start_index = 1 ;
        double node_lat(node) ;
                node_lat:standard_name = "latitude" ;
                node_lat:long_name = "Latitude of 2D mesh nodes" ;
                node_lat:units = "degrees_north" ;
        double node_lon(node) ;
                node_lon:standard_name = "longitude" ;
                node_lon:long_name = "Longitude of 2D mesh nodes" ;
                node_lon:units = "degrees_east" ;
        float surface_temp(node) ;
                surface_temp:coordinates = "node_lon node_lat" ;
                surface_temp:long_name = "Sea surface temperature" ;
                surface_temp:standard_name = "sea_surface_temperature" ;
                surface_temp:mesh = "Mesh" ;
                surface_temp:location = "node" ;
                surface_temp:units = "degree_Celsius" ;

// global attributes:
                :Conventions = "CF-1.6/UGRID-1.x" ;

data:

 Mesh = _ ;

 face_nodes =
  1, 4, 5,
  1, 2, 3,
  1, 3, 4;

 node_lat = 40, 50,
         50, 40, 30;

 node_lon = 0, 10,
         20, 30, 10;

 surface_temp = 200, 100, 0, 50, 150;

}
```
![image](https://user-images.githubusercontent.com/5402758/223718756-f965951b-ec09-4191-b57c-f3c0d709696c.png)

But I am not able to reduce to only 4 or 3 nodes. Do not understand why ?
I am also looking for testing face values (not on nodes).

**Reply by lesserwhirls** [reply]
I will need some free time on the weekend to dig into why exactly you are unable to reduce your example dataset (not sure when I will have the time though). In the meantime, have a look at [this](https://www.icloud.com/iclouddrive/0d6OYMU89J5pijcSXCfSm4YlA#usf_fvcom_2019_01_28) netCDF file, which works with the TDS and edal-java (provides WMS services) without needing any NcML modifications. This example file was produced by fvcom over the west Florida shelf and includes both node and face variables.

**Reply by PBrockmann** [reply]
Thank you Sean.
What I am missing is very simple examples.
Here working on how to represent 3 triangles with face values.
I am looking into your file. I should be ok with that. I will let you know.

**Reply by PBrockmann** [reply]
Nope. Even by inspecting the usf_fvcom_2019_01_28.nc that by the way do not respect the UGRID convention when I use the 
ugrid-checker

```
$ ugrid-checker usf_fvcom_2019_01_28.nc
UGRID conformance checks complete.

List of checker messages :
  *** FAIL R202 : Mesh coordinate variable "lonc" within fvcom_mesh:face_coordinates has dimension "nele", but the parent mesh face dimension is "three".
  *** FAIL R202 : Mesh coordinate variable "latc" within fvcom_mesh:face_coordinates has dimension "nele", but the parent mesh face dimension is "three".
  *** FAIL R504 : Mesh data variable "hc" has location="cell", which is not one of "face", "edge" or "node".
  *** FAIL R509 : Mesh data variable "hc" has dimensions ('nele',), of which 0 are mesh dimensions, instead of 1.
  *** FAIL R509 : Mesh data variable "u" has dimensions ('time', 'siglay', 'nele'), of which 0 are mesh dimensions, instead of 1.
  *** FAIL R509 : Mesh data variable "v" has dimensions ('time', 'siglay', 'nele'), of which 0 are mesh dimensions, instead of 1.

Total of 6 problems logged :
  6 Rxxx requirement failures
  0 Axxx advisory recommendation warnings

Done.
```

So I don't get it. Here is a simple file with 3 triangles that I cannot plot with the WMS.
I get triangles but cut.

![image](https://user-images.githubusercontent.com/5402758/223803901-ac6a24f9-2ed2-440c-a4ac-85c37ca78570.png)

Here is the cdl.

```
netcdf ugrid9 {
dimensions:
        face = 3 ;
        node = 5 ;
        three = 3 ;
variables:
        int Mesh ;
                Mesh:cf_role = "mesh_topology" ;
                Mesh:long_name = "Topology data of 2D unstructured mesh" ;
                Mesh:topology_dimension = 2 ;
                Mesh:node_coordinates = "node_lat node_lon" ;
                Mesh:face_node_connectivity = "face_nodes" ;
                Mesh:face_coordinates = "lat_c lon_c" ;
        int face_nodes(three, face) ;
                face_nodes:cf_role = "face_node_connectivity" ;
                face_nodes:long_name = "Maps every face to its corner nodes." ;
                face_nodes:start_index = 1 ;
        double node_lat(node) ;
                node_lat:standard_name = "latitude" ;
                node_lat:long_name = "Latitude of 2D mesh nodes" ;
                node_lat:units = "degrees_north" ;
        double node_lon(node) ;
                node_lon:standard_name = "longitude" ;
                node_lon:long_name = "Longitude of 2D mesh nodes" ;
                node_lon:units = "degrees_east" ;
        double lat_c(face) ;
                lat_c:standard_name = "latitude" ;
                lat_c:long_name = "Latitude of 2D mesh nodes" ;
                lat_c:units = "degrees_north" ;
        double lon_c(face) ;
                lon_c:standard_name = "longitude" ;
                lon_c:long_name = "Longitude of 2D mesh nodes" ;
                lon_c:units = "degrees_east" ;
        float surface_temp(face) ;
                surface_temp:coordinates = "node_lon_c node_lat_c" ;
                surface_temp:long_name = "Sea surface temperature" ;
                surface_temp:standard_name = "sea_surface_temperature" ;
                surface_temp:mesh = "Mesh" ;
                surface_temp:location = "face" ;
                surface_temp:units = "degree_Celsius" ;

// global attributes:
                :Conventions = "CF-1.6/UGRID-1.0" ;

data:

 Mesh = _ ;

 face_nodes =
  1, 5, 4,
  1, 3, 2,
  1, 4, 3;

 node_lat = 40, 50, 50, 40, 30;
 node_lon = 0, 10, 20, 30, 10;
 lat_c = 35, 45, 45;
 lon_c = 20, 7.5, 15;

 surface_temp = 200, 100, 0;

}
```

I do not understand what is uncorrect. Do the centers of the faces need to be at the centroïds ?

**Reply by PBrockmann** [reply]
A simplified version of your file usf_fvcom_2019_01_28.nc gives https://thredds-su.ipsl.fr/thredds/fileServer/ipsl_thredds/brocksce/tmp/usf_fvcom_2019_01_28_extract2.nc (only one variable on faces).
and this map:

![image](https://user-images.githubusercontent.com/5402758/223847943-29246791-a4db-48d6-b7f1-9c4e5a31a03a.png)

What is odd is that it also gives problems with the ugrid-checker (https://pypi.org/project/ugrid-checks/)

```
$ ugrid-checker usf_fvcom_2019_01_28_extract2.nc

UGRID conformance checks complete.

List of checker messages :
  *** FAIL R202 : Mesh coordinate variable "lonc" within fvcom_mesh:face_coordinates has dimension "nele", but the parent mesh face dimension is "three".
  *** FAIL R202 : Mesh coordinate variable "latc" within fvcom_mesh:face_coordinates has dimension "nele", but the parent mesh face dimension is "three".
  *** FAIL R509 : Mesh data variable "u" has dimensions ('nele',), of which 0 are mesh dimensions, instead of 1.

Total of 3 problems logged :
  3 Rxxx requirement failures
  0 Axxx advisory recommendation warnings

Done.
```

Error disappears if you change face_node_connectivity dimension order:
`int nv(three, nele)` to `int nv(nele, three)` but then the map is not correct.
The ugrid-checks tool may not be agreed to the reading from the ncWMS UGRID. 
See also https://github.com/ugrid-conventions/ugrid-conventions/issues/57

**Reply by PBrockmann** [reply]
What would really help me is to have two single triangle UGRID netcdf files. One with node values (3) and one a face value (1).

**Comment by PBrockmann** [comment]
@lesserwhirls you wrote: "this implementation only handles triangular meshes"
So hexaedron cannot be represented ?
That might be a problem for representing DYNAMICO output that uses an icosahedral grid.

**Comment by PBrockmann** [comment]
Any news on representing a single triangle with a face value ?
This simple representation would be help a lot to understand how TDS can represent a UGRID variable.

**Comment by lesserwhirls** [comment]
I've finally had some spare time to dig into this a bit. For a single triangle (three nodes, one face), the following cdl should generate an ok netCDF file:

~~~
netcdf bermuda_triangle {
dimensions:
        node = 3 ;
        face = 1 ;
        three = 3 ;
        time = 1 ;
variables:
        int fvcom_mesh ;
                fvcom_mesh:cf_role = "mesh_topology" ;
                fvcom_mesh:topology_dimension = 2 ;
                fvcom_mesh:node_coordinates = "lon lat" ;
                fvcom_mesh:face_node_connectivity = "nv" ;
        int time(time) ;
                time:units = "seconds since 1970-01-01" ;
                time:standard_name = "time" ;
        double lat(node) ;
                lat:units = "degrees_north" ;
        double latc(face) ;
                latc:units = "degrees_north" ;
        double lon(node) ;
                lon:units = "degrees_east" ;
        double lonc(face) ;
                lonc:units = "degrees_east" ;
        int nv(face, three) ;
                nv:long_name = "nodes surrounding element" ;
                nv:cf_role = "face_node_connectivity" ;
                nv:start_index = 0 ;
        double h(node) ;
                h:coordinates = "lat lon" ;
                h:mesh = "fvcom_mesh" ;
                h:location = "node" ;
                h:units = "m" ;
        float temp(time, node) ;
                temp:coordinates = "time lat lon" ;
                temp:mesh = "fvcom_mesh" ;
                temp:location = "node" ;
                temp:units = "degree_Celsius" ;
        float u(time, face) ;
                u:coordinates = "time latc lonc" ;
                u:mesh = "fvcom_mesh" ;
                u:location = "face" ;
                u:units = "m s-1" ;

// global attributes:
                :Conventions = "CF-1.X/UGRID-1.X" ;
data:

 fvcom_mesh = _ ;

 time = 1681674619 ;

 lat = 27.789, 18.466, 32.295 ;

 latc = 25.52 ;

 lon = -80.227, -66.106, -64.781 ;

 lonc = -70.37 ;

 nv =
  0, 1, 2 ;

 h = 9.1, 10.2, 11.3 ;

 temp =
  20.1, 21.2, 23.3 ;

 u =
  -0.0775 ;
}
~~~

However, if I am reading the edal-java code correctly, there is a bad assumption being made here:

https://github.com/Reading-eScience-Centre/edal-java/blob/e881f5deb8d0c219f305418ba07803e1546a1cb4/cdm/src/main/java/uk/ac/rdg/resc/edal/dataset/cdm/CdmGridDatasetFactory.java#L837-L859

For triangular meshes, the `face_node_connecitivity` variable should have dimensionality `number_of_faces, 3`. In the case of a single triangle (one face, each face with three edges), the edal-java code will incorrectly swap dimensions and throw errors. That's likely why you cannot go below four nodes (as two faces would trigger the incorrect behavior). I'd suggest filing a bug with the `edal-java` project.

**Reply by PBrockmann** [reply]
Thank you very much.
I will test this.

Is there any chance to have a hexahedron be represented ?

**Reply by lesserwhirls** [reply]
You would need to work with the `edal-java` project regarding hexahedron meshes in WMS, as their code would need to be extended.

**Reply by PBrockmann** [reply]
No sucess when I try to display the bermuda_triangle.
Nothing visible.

**Reply by lesserwhirls** [reply]
Right - there is a bug in `edal-java` that is preventing the single triangle case from working. See the last part of https://github.com/Unidata/tds/discussions/345#discussioncomment-5630314.

**Reply by PBrockmann** [reply]
Yes, it should be a good thing to describe this bug in an issue on the edal-java project
But there is already one about the subject - see https://github.com/Reading-eScience-Centre/edal-java/issues/121, but still open since Oct 2019...
